tools/integration_tests/operations/parallel_dirops_test.go

```diff
@@ -0,0 +1,447 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A collection of E2E tests for a file system where parallel dirops are allowed.
+// Dirops refers to readdir and lookup operations. These tests are complimentary
+// to the composite tests.
+package operations_test
+
+import (
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/tools/integration_tests/util/operations"
+	"github.com/googlecloudplatform/gcsfuse/v2/tools/integration_tests/util/setup"
+	"github.com/stretchr/testify/assert"
+)
+
+// createDirectoryStructureForParallelDiropsTest creates the following files and
+// directory structure.
+// bucket
+//
+//			file1.txt
+//			file2.txt
+//			explicitDir1
+//					file1.txt
+//					file2.txt
+//	    explicitDir2
+//					file1.txt
+func createDirectoryStructureForParallelDiropsTest(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	setup.CleanUpDir(testDir)
+
+	// Create explicitDir1 structure
+	explicitDir1 := path.Join(testDir, "explicitDir1")
+	operations.CreateDirectory(explicitDir1, t)
+	filePath1 := path.Join(explicitDir1, "file1.txt")
+	operations.CreateFileOfSize(5, filePath1, t)
+	filePath2 := path.Join(explicitDir1, "file2.txt")
+	operations.CreateFileOfSize(10, filePath2, t)
+
+	// Create explicitDir2 structure
+	explicitDir2 := path.Join(testDir, "explicitDir2")
+	operations.CreateDirectory(explicitDir2, t)
+	filePath1 = path.Join(explicitDir2, "file1.txt")
+	operations.CreateFileOfSize(11, filePath1, t)
+
+	filePath1 = path.Join(testDir, "file1.txt")
+	operations.CreateFileOfSize(5, filePath1, t)
+	filePath2 = path.Join(testDir, "file2.txt")
+	operations.CreateFileOfSize(3, filePath2, t)
+}
+
+func TestParallelLookUpsForSameFile(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	// Create directory structure for testing.
+	createDirectoryStructureForParallelDiropsTest(t)
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
```

Comment:
That's a good point, however I see that only lookUp is used 6 times and rest are used 1-2 times. So I think it's okay to have this duplication. Lmk if you still think otherwise and i will make changes.

Reply:
I think either way is okay, approving! :)